src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalEngine.java

```diff
@@ -0,0 +1,922 @@
+
+    // The set of all possible evaluation contexts
+    private StratificationManager<VariantStratifier, EvaluationContext> stratManager;
+
+    private SampleDB sampleDB = null;
+
+    private List<FeatureInput<VariantContext>> knowns = new ArrayList<>();
+
+    // maintain the mapping of FeatureInput to name used in output file
+    private Map<FeatureInput<VariantContext>, String> inputToNameMap = new HashMap<>();
+
+    // Variables
+    private Set<SortableJexlVCMatchExp> jexlExpressions = new TreeSet<>();
+
+    private boolean isSubsettingSamples;
+    private Set<String> sampleNamesForEvaluation = new LinkedHashSet<>();
+    private Set<String> familyNamesForEvaluation = new LinkedHashSet<>();
+    private Set<String> sampleNamesForStratification = new LinkedHashSet<>();
+    private Set<String> familyNamesForStratification = new LinkedHashSet<>();
+
+    // important stratifications
+    private boolean byFilterIsEnabled = false;
+    private boolean perSampleIsEnabled = false;
+    private boolean perFamilyIsEnabled = false;
+
+    private AlleleFrequency.StratifyingScale AFScale = AlleleFrequency.StratifyingScale.LINEAR;
+    private boolean useCompAFStratifier = false;
+
+    // maintain the mapping of source name (from VC) to FeatureInput name
+    private Map<String, FeatureInput<VariantContext>> drivingVariantSourceMap;
+
+    // No args constructor for unit testing only
+    @VisibleForTesting
+    protected VariantEvalEngine() {
+        this.variantEvalArgs = new VariantEvalArgumentCollection();
+        this.samSequenceDictionaryForDrivingVariants = null;
+        this.traversalIntervals = null;
+        this.features = null;
+    }
+
+    public VariantEvalEngine(final VariantEvalArgumentCollection variantEvalArgs, final FeatureManager features, final List<SimpleInterval> traversalIntervals, SAMSequenceDictionary samSequenceDictionaryForDrivingVariants, @Nullable final Collection<String> samples) {
+        this(variantEvalArgs, features, traversalIntervals, samSequenceDictionaryForDrivingVariants, samples, false);
+    }
+
+    // Note: this protected constructor with deferred initialization exists for DISCVR-seq/VariantQC
+    protected VariantEvalEngine(VariantEvalArgumentCollection variantEvalArgs, FeatureManager features, List<SimpleInterval> traversalIntervals, SAMSequenceDictionary samSequenceDictionaryForDrivingVariants, @Nullable Collection<String> samples, boolean skipInitialize) {
+        this.variantEvalArgs = variantEvalArgs;
+        this.samSequenceDictionaryForDrivingVariants = samSequenceDictionaryForDrivingVariants;
+        this.traversalIntervals = traversalIntervals;
+        this.features = features;
+
+        // Cache map of source name -> FeatureInput
+        drivingVariantSourceMap = new HashMap<>();
+        variantEvalArgs.getFeatureInputsForDrivingVariants().forEach(x -> drivingVariantSourceMap.put(x.getName(), x));
+
+        if (!skipInitialize) {
```

Comment:
I thought we were getting close, but this deferred initialization state seems complicating - whats the reason for needing it (both here, and in the evaluators, which still have an "initialize" method) ? Is there any alternative ?

Depending on the answer, we'll need to rationalize these constructors - having one that controls initialization and one that doesn't seems problematic. There are 3 constructors now, each of which leaves the object in a slightly different non-obvious state. We really need to get rid of the boolean if at all possible.

Reply:
The reason for this is listed in the comment above that method: it was to make this work with our VariantQC tool. I need to delay creation of EvaluationContext for some things our tool does. However, I took a look at my code and I think I have a solution that doesnt need this, so I removed it here.